mlir/tools/rocmlir-gen/rocmlir-gen.cpp

```diff
@@ -2232,10 +2317,19 @@ static func::FuncOp createVerifierFunc(ModuleOp module, const KernelIF &kernel,
     // %5 = memref.cast %0 : memref<802816xf32> to memref<?x?x?x?x?xf32>
     // clang-format on
 
-    testResultNew = b.create<memref::AllocOp>(loc, valFlatType);
+    auto f32FlatType = MemRefType::get(valFlatType.getShape(), floatType);
+    testResultNew = b.create<memref::AllocOp>(loc, f32FlatType);
     emitMemcpy(b, testFlat, testResultNew);
-    testResult = b.create<memref::CastOp>(loc, mr1DUnkValType, testResultNew);
-    mr1DUnkTestType = mr1DUnkValType;
+    testResult = b.create<memref::CastOp>(loc, mr1DUnkF32Type, testResultNew);
+    mr1DUnkTestType = mr1DUnkF32Type;
+    if (!valElemType.isF32()) {
```

Comment:
Treat the validation data exactly like the test-result data and promote to F32 before calling mcpuVerify.